### ISSUE TITLE
sockjs: WebSocketSession not yet initialized

### ISSUE THREAD

**Issue by elijahiuu** [chimurai/http-proxy-middleware#57]
I am having problems proxying SockJS websockets.  On the server side I am getting "WebSocketSession not yet initialized". On the client side I am getting "failed: Invalid frame header". It looks like it almost connects, but something goes wrong in handshake. Any ideas?

Here's my proxy configuration.

```javascript
var gulp            = require('gulp');
var express         = require('express');
var proxyMiddleware = require('http-proxy-middleware');

gulp.task('serve', function() {
    proxy = proxyMiddleware('/go', {
        target: 'http://localhost:8080/',
        ws: true,
    });

    var app = express();
    app.use('/', express.static('./build/'));             
    app.use(proxy);                                       

    var server = app.listen(3000);
    server.on('upgrade', proxy.upgrade);                  
});

gulp.task('build', function(){
    gulp.src('src/**/*.*')
        .pipe(gulp.dest('./build/go/'));
});
```

Network tab:
**General**
```
Request URL:ws://localhost:3000/go/xxxxxxxxxx/websocket
Request Method:GET
Status Code:101 Switching Protocols
```
**Response Headers**
```
view source
connection:upgrade
date:Sun, 20 Mar 2016 20:59:11 GMT
sec-websocket-accept:sPIVWZeoUMNdIxL4EHnYgvBoZng=
sec-websocket-extensions:permessage-deflate;client_max_window_bits=15
server:Apache-Coyote/1.1
upgrade:websocket
```

**Comment by chimurai** [comment]
If your server is virtual hosted, try it with the option: `changeOrigin: true`

```javascript
    proxy = proxyMiddleware('/go', {
        target: 'http://localhost:8080/',
        ws: true,
        changeOrigin: true
    });

```

**Comment by elijahiuu** [comment]
It is a local Tomcat running a java app and hosting the socket. I tried `changeOrigin`, that gave me: 

``` 
WebSocket connection to 'ws://localhost:3000/go/xxxxxxxxx/websocket' failed: Connection closed before receiving a handshake response 
```

**Comment by chimurai** [comment]
Maybe you can leave out `ws:true` from the options or set it to `ws:false`, since you are explicitly upgrading with `server.on('upgrade', proxy.upgrade);`

**Comment by elijahiuu** [comment]
I think that actually works! Thank you @chimurai

**Comment by chimurai** [comment]
Glad to hear the solution works! :)

**Comment by chimurai** [comment]
@elijahiuu Now that I think about it, this might be a bug...

I'll have to spend some time to investigate this.

**Comment by sashadt** [comment]
+1  Having same problem, and seeing:

```
WebSocket connection to 'ws://localhost:9000/pubsub' failed: Invalid frame header
```
removing the line `ws: true` from proxy options seems to fix it.

I agree it is either a bug, or usage documentation needs to be updated.

**Comment by chimurai** [comment]
@sashadt @elijahiuu 

Think I've found the issue...

I published a beta version: v0.15.1-beta.
You can get it with: `npm install http-proxy-middleware@0.15.1-beta`

Can you test it to confirm the fix?

Problem shouldn't occur anymore when the following are used:
```javascript
// proxy config
config.ws: true

// external upgrade
server.on('upgrade', proxy.upgrade); 
```
